include/ginkgo/core/distributed/row_gatherer.hpp

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_ROW_GATHERER_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_ROW_GATHERER_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <future>
+
+#include <ginkgo/core/base/dense_cache.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/collective_communicator.hpp>
+#include <ginkgo/core/distributed/index_map.hpp>
+#include <ginkgo/core/distributed/lin_op.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace distributed {
+
+
+/**
+ * The distributed::RowGatherer gathers the rows of distributed::Vector that
+ * are located on other processes.
+ *
+ * Example usage:
+ * ```c++
+ * auto coll_comm = std::make_shared<mpi::neighborhood_communicator>(comm,
+ * imap); auto rg = distributed::RowGatherer<int32>::create(exec, coll_comm,
+ * imap);
+ *
+ * auto b = distributed::Vector<double>::create(...);
+ * auto x = matrix::Dense<double>::create(...);
+ *
+ * auto future = rg->apply_async(b, x);
+ * // do some computation that doesn't modify b, or access x
```

Comment:
I guess this just meant to say that you can't expect any meaningful data when accessing `x` before the `wait` has completed.

Reply:
I think I get it wrong.
Is the comment here to describe that user can do something safely after the call or the apply_async behavior?
My comment was based on that it is the behavior of the apply_async because apply_async definitely accesses x.
If it is for user action during async and wait, then it is correct.